### ISSUE TITLE
@urql/exchange-graphcache Memory leak if using makeDefaultStorage

### ISSUE THREAD

**Issue by Mookiies** [FormidableLabs/urql#1876]
When using `makeDefaultStorage` there is a memory leak if clients are no longer in use. This is because the listener from `onOnline` is never removed.

This could also be dangerous because since old storage is still subscribed to the event listener it will have it's callbacks called. I haven't really looked into this.

Ideally there would be some sort of API for unsubscribing these listeners in addition to not having this leak for `makeDefaultStorage`. Which would make it easy for custom implementations of storage to not have to micro-manage the listeners that are required by storage.

If you create many re-create client many times you can see the memory footprint continue to climb. Using a production build of react (https://mookiies.github.io/onOnlineDemo/) memory footprint went from 1.7mb to 7.6mb after creating a client 150 times.

urql version & exchanges:
` "@urql/exchange-graphcache": "^4.3.0",`
`"urql": "^2.0.4",`

**Steps to reproduce**

1. Create a client that uses `makeDefaultStorage`
2. Destroy client 

**Expected behavior**
- Client instances are garbage collected 

**Actual behavior**
- Client instances are kept around

### Code Example
Based off the example of destroying client for cache invalidation here: https://formidable.com/open-source/urql/docs/advanced/authentication/#cache-invalidation-on-logout
 
<details>
  <summary>Click here to expand</summary>

```
import React, { useState, useMemo } from "react";
import {
  createClient,
  dedupExchange,
  fetchExchange,
  Provider
} from "urql";

import { offlineExchange } from "@urql/exchange-graphcache";
import { makeDefaultStorage } from "@urql/exchange-graphcache/default-storage";
import { Todos } from "./components";

const storage = makeDefaultStorage({
  idbName: "graphcache-v3", // The name of the IndexedDB database
  maxAge: 7 // The maximum age of the persisted data in days
});

const cache = offlineExchange({
  storage
});

const clientConfig = {
  url: "https://0ufyz-4000.sse.codesandbox.io",
  exchanges: [
    dedupExchange,
    cache,
    fetchExchange
  ],
  requestPolicy: "cache-first"
};

const App = () => {
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const toggleLoggedIn = () => {
    setIsLoggedIn((previousValue) => !previousValue);
  };
  const client = useMemo(() => {
    if (!isLoggedIn) {
      return null;
    }

    console.log('createClient');
    return createClient(clientConfig);
  }, [isLoggedIn]);

  const app = <Provider value={client}>
     <Todos />
  </Provider>

  return (
      <main>
        <button onClick={toggleLoggedIn}>Toggle login</button>
        {isLoggedIn && app}
      </main>
  );
};

export default App;
```
</details>

Demo deployed here with the above code that can be profiled, but easier to run locally with unminified js
client: https://mookiies.github.io/onOnlineDemo/ 
server: https://0ufyz-4000.sse.codesandbox.io
(built from https://github.com/Mookiies/onOnlineDemo)

**Comment by kitten** [comment]
We did realise that, but overall the client doesn't have a lot of mechanisms to be exposed, since they're not commonly needed. Granted, we could dispose of the storage and allow this to be done manually. However, you're saying this gets severe when you create 150 clients, right? I find this to be a little of an unlikely scenario, so I'm wondering how relevant this actually is, given that it adds more user requirements.

I'm not super opposed to adding disposers to storage, but I am at the same time curious whether this is worth solving 😅 Our last conclusion was that it's unlikely to cause any issues for prolonged times/sessions.

**Comment by Mookiies** [comment]
I agree that the memory footprint side probably isn't the biggest deal as it's unlikely that a client will re-created _that_ many times.

What I'm really more concerned about is the potential side effects that could arise from having a client that is no longer being used trying to execute mutations and potentially overriding cache items as a result.  Are there any mechanism that would stop and old client from trying to execute?

**Comment by kitten** [comment]
Oh, I see, i.e. in a scenario where it's discarded, comes back online, and reruns several mutations? Hm, we did consider that but we assumed that since the "dispatcher" of these mutations couldn't tell whether the client was online or offline, it also shouldn't matter too much whether those are eventually sent or not 🤔
